### PR TITLE
Support data for Bidi properties with combined data structure

### LINKED ISSUE #2833: Provide lookup for the Bidi_Mirroring_Glyph property
[HarfBuzz wants to be able to look up](https://searchfox.org/mozilla-central/rev/8cfe1317c14fe6c997193141b8807d58f92c4f48/gfx/harfbuzz/src/hb-unicode.h#597-611) the [`Bidi_Mirroring_Glyph` property](https://www.unicode.org/Public/UCD/latest/ucd/BidiMirroring.txt) for a character.

To be able to use ICU4X as a HarfBuzz Unicode data back end, ICU4X should provide lookup for this property.

### PR DESCRIPTION
**Pull request by echeran**
Fixes #2833

### CODE REVIEW

File: components/properties/src/props.rs [hunk 1]
```diff
@@ -870,3 +879,31 @@ impl CanonicalCombiningClass {
...
+///
+/// Also see `u_charMirror` in ICU4C.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "datagen", derive(databake::Bake))]
+#[cfg_attr(feature = "datagen", databake(path = icu_properties))]
+#[allow(clippy::exhaustive_structs)] // newtype
+#[repr(transparent)]
+#[zerovec::make_ule(BidiMirroringGlyphULE)]
+pub struct BidiMirroringGlyph(pub u32);
```
Comment (sffc) [inline components/properties/src/props.rs:909]: Issue: We should make this a 3-byte type so that the data provider also stores it as 3 bytes.

`[u8; 3]` is one way to accomplish this. Alternatively, a cleaner approach would be to add `UnvalidatedChar` in the zerovec crate with the correct semantics.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/properties/src/props.rs (hunk 1 above)
Comment (hsivonen) [reply components/properties/src/props.rs:909]: I'd like to use the high 3 bits to fit more related properties in this data structure, so just "unvalidated char" wouldn't the the whole semantic. We might reuse the [`U24` struct](https://github.com/unicode-org/icu4x/commit/90db7c6a5ef3d64e969850c2655e8b9e45948c81#diff-8d43906dd2fed874a89265d4d6b07aa4f04c5a391fedbc66fb4b912251ed70b4) that the normalizer had before.

Quoting self from Slack:

> I do think it makes sense to bundle at least Bidi_Mirroring_Glyph, Bidi_Mirrored, Bidi_Paired_Bracket , and Bidi_Paired_Bracket_Type as a 24-bit value in one trie. I'm less sure of whether it makes sense to extend the trie value to 32 bits and bake Bidi_Class in the additional 8 bits.

File: components/properties/src/bidi_data.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. MirroredPairedBracketDataULE does not include any uninitialized or padding bytes
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  2. MirroredPairedBracketDataULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. CharULE byte equality is semantic equality
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:45]:
```suggestion
//  6. MirroredPairedBracketDataULE byte equality is semantic equality
```

File: components/properties/src/bidi_data.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+/// This module provides an efficient storage of data serving the following
+/// properties:
+/// - `Bidi_Paired_Bracket`
+/// - `Bidi_Paired_Bracket_Type`
+/// - `Bidi_Mirrored`
+/// - `Bidi_Mirroring_Glyph`
+///
+
+
+/// 20..0  Code point return value for Bidi_Mirroring_Glyph value
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:21]: these docs should go on the ULE type

Also, this module should probably be a submodule of `provider` and marked unstable the way we usually do provider code, but not `doc(hidden)`

Comment (Manishearth) [reply components/properties/src/bidi_data.rs:21]: also these ranges seem backwards?

File: components/properties/src/bidi_data.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            let _mirroring_glyph = char::from_u32(mirroring_glyph_code_point)
+                .ok_or(ZeroVecError::VarZeroVecFormatError)?;
+
+            // skip validating the Bidi_Mirrored boolean since it is always valid
+
+            // skip validating Bidi_Paired_Bracket_Type because enum is open
+            //
+            // TODO: is the above statement correct and okay?
+            // TODO: or do we have a way to validate the enum discriminant against the enum for
+            //       newtype idiom enums?
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:75]: issue: we must validate that all _unused_ bytes are 0 (or some other fixed value)

File: components/properties/src/bidi_data.rs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        let mirroring_glyph_ule_bytes = &[unaligned.0[0], unaligned.0[1], unaligned.0[2] | 0x1F];
+        // Safe because the lower bits 20..0 of MirroredPairedBracketDataULE bytes are the CharULE bytes,
+        // and CharULE::from_unaligned is safe because bytes are defined to represent a valid Unicode code point.
+        let mirroring_glyph = unsafe {
+            let mirroring_glyph_ule = CharULE::from_byte_slice_unchecked(mirroring_glyph_ule_bytes);
+            char::from_unaligned(mirroring_glyph_ule[0])
+        };
+        let is_mirrored = unaligned.0[2] | 0x20 == 1;
+        let paired_bracket_type =
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:106]: just use `BidiPairedBracketType(integer)`, no need to go via the ULE type since we're just dealing in bytes

Comment (Manishearth) [reply components/properties/src/bidi_data.rs:106]: issue: all of the bitwise math should be documented with main docs on the ULE type itself that lists all the ways we mask things

File: components/properties/src/bidi_data.rs [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            BidiPairedBracketType::from_unaligned(BidiPairedBracketTypeULE(unaligned.0[2] | 0xC0));
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:107]: issue: this mask is not correct, given that it will produce values greater than 2.

File: components/properties/src/bidi_data.rs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    /// let close_paren = bidi_data.get_mirroring_props(')' as u32);
+    /// assert_eq!(close_paren, BidiMirroringProperties{ mirroring_glyph: Some('('), mirrored: true });
+    /// let open_angle_bracket = bidi_data.get_mirroring_props('<' as u32);
+    /// assert_eq!(open_angle_bracket, BidiMirroringProperties{ mirroring_glyph: Some('>'), mirrored: true });
+    /// let close_angle_bracket = bidi_data.get_mirroring_props('>' as u32);
+    /// assert_eq!(close_angle_bracket, BidiMirroringProperties{ mirroring_glyph: Some('<'), mirrored: true });
+    /// let three = bidi_data.get_mirroring_props('3' as u32);
+    /// assert_eq!(three, BidiMirroringProperties{ mirroring_glyph: None, mirrored: false });
+    /// ```
+    pub fn get_mirroring_props(&self, code_point: u32) -> BidiMirroringProperties {
```
Comment (sffc) [inline components/properties/src/bidi_data.rs:113]: Nit: since it takes a u32 this function should be called `get32_mirroring_props`; you could add another function that takes a `char` instead.

File: components/properties/src/props.rs [hunk 8]
```diff
@@ -284,6 +289,26 @@ impl BidiClass {
     pub const PopDirectionalIsolate: BidiClass = BidiClass(22);
 }
 
+/// TODO: documentation for Enumerated property Bidi_Paired_Bracket_Type
```
Comment (sffc) [inline components/properties/src/props.rs:292]: Please fix

File: components/properties/src/provider/bidi_data.rs [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+        let err_msg = "Could not deserialize MirroredPairedBracketData from u32";
+        let byte0 = bytes
+            .first()
+            .ok_or(ZeroVecError::ParseError { ty: err_msg })?;
+        let byte1 = bytes
+            .get(1)
+            .ok_or(ZeroVecError::ParseError { ty: err_msg })?;
+        let byte2 = bytes
+            .get(2)
+            .ok_or(ZeroVecError::ParseError { ty: err_msg })?;
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:107]: Suggestion:

```suggestion
        let[byte0, byte1, byte2, byte3] = u32::to_le_bytes(x);
        if byte3 != 0 {
            return Err(ZeroVecError::parse::<Self>());
        }
```

File: components/properties/src/provider/bidi_data.rs [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        let ule_byte_array = [*byte0, *byte1, *byte2];
+        let ule_struct_slice = MirroredPairedBracketDataULE::parse_byte_slice(&ule_byte_array)?;
+        let ule_struct = ule_struct_slice
+            .first()
+            .ok_or(ZeroVecError::ParseError { ty: err_msg })?;
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:112]: Nit:

```suggestion
            .ok_or(ZeroVecError::parse::<Self>())?;
```

File: components/properties/src/provider/bidi_data.rs [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+            byte_slice.first().copied().unwrap_or_default(),
+            byte_slice.get(1).copied().unwrap_or_default(),
+            byte2,
+        ])
+    }
+
+    #[inline]
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        // Safe because the lower bits 20..0 of MirroredPairedBracketDataULE bytes are the CharULE bytes,
+        // and CharULE::from_unaligned is safe because bytes are defined to represent a valid Unicode code point.
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:204]: Nit: Only wrap the single unsafe expression in the unsafe block

File: components/properties/src/provider/bidi_data.rs [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+        let mirroring_glyph = unsafe {
+            let mirroring_glyph_ule_bytes =
+                &[unaligned.0[0], unaligned.0[1], unaligned.0[2] & 0x1F];
+            let mirroring_glyph_ule = CharULE::from_byte_slice_unchecked(mirroring_glyph_ule_bytes);
+            mirroring_glyph_ule
+                .first()
+                .map(|ule| char::from_unaligned(*ule))
+                .unwrap_or(char::REPLACEMENT_CHARACTER)
+        };
+        let mirrored = (unaligned.0[2] >> 5) & 0x1 == 1;
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:214]: Nit (optional): I don't like having to look up operator precedence

```suggestion
        let mirrored = ((unaligned.0[2] >> 5) & 0x1) == 1;
```

File: components/properties/src/props.rs [hunk 13]
```diff
@@ -284,6 +289,26 @@ impl BidiClass {
...
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "datagen", derive(databake::Bake))]
+#[cfg_attr(feature = "datagen", databake(path = icu_properties))]
+#[allow(clippy::exhaustive_structs)] // newtype
+#[repr(transparent)]
+#[zerovec::make_ule(BidiPairedBracketTypeULE)]
```
Comment (sffc) [inline components/properties/src/props.rs:299]: Nit: I think this does not need to be a ULE. Let's not add ULE impls unless we need them.

File: components/properties/src/trievalue.rs [hunk 14]
```diff
@@ -163,3 +173,22 @@ impl TrieValue for GeneralCategoryGroup {
         u16::try_from(i).map(packed_u16_to_gcg)
     }
 }
+
+impl TrieValue for MirroredPairedBracketData {
+    type TryFromU32Error = CharTryFromError;
+
+    fn try_from_u32(i: u32) -> Result<Self, Self::TryFromU32Error> {
```
Comment (sffc) [inline components/properties/src/trievalue.rs:180]: Nit/Suggestion: you already have `impl TryFrom<u32> for MirroredPairedBracketData` and you should call it from here.

File: components/properties/src/trievalue.rs [hunk 15]
```diff
@@ -97,6 +99,14 @@ impl TrieValue for SentenceBreak {
     }
 }
 
+impl TrieValue for BidiPairedBracketType {
```
Comment (sffc) [inline components/properties/src/trievalue.rs:102]: Nit: Don't add these impls unless we need them for something, and I _think_ we don't need it for this one

File: provider/repodata/data/json/props/bidiauxiliaryprops@1/und.json [hunk 16]
```diff
@@ -0,0 +1,6633 @@
...
+      {
+        "mirroring_glyph": "\u0000",
+        "mirrored": false,
+        "paired_bracket_type": 0
+      },
+      {
+        "mirroring_glyph": "\u0000",
+        "mirrored": false,
+        "paired_bracket_type": 0
+      },
```
Comment (sffc) [inline provider/repodata/data/json/props/bidiauxiliaryprops@1/und.json:1615]: Observation: the great majority of these entries are all the same. It would be nice if CPT gave us better compression.

**Review by sffc — COMMENTED** [review]
Really good stuff! This looks basically ready to go to me, modulo a number of minor things.

I did not look too carefully at the datagen code; I focused on the component code.

File: provider/repodata/data/json/props/bidiauxiliaryprops@1/und.json (hunk 16 above)
Comment (echeran) [reply provider/repodata/data/json/props/bidiauxiliaryprops@1/und.json:1615]: CPT Builder compaction isn't a compression algorithm like gzip. It just tries to find ways in which the value in the last hop from the index array to the data array can be reused without loss of correctness. The last 6 bits of the code point's bit field are used to advance from the offset index in the data array given by the index array. The 2^6=64 values starting from that offset are called a "block", and so that's the level of granularity (more or less) that compaction can happen).

The levels of the trie depend on the number and size of partitions of the code point's bit field, and those decisions all a tradeoff between space and time. The ones for this iteration of CodePointTrie also work well for UTF-8 encodings (ex: the 6-bit size allows you to skip some of the work of reconstructing the code point from the UTF-8 trail bytes). I don't know if that answered your question, but hopefully it was relevant.

@markusicu

File: components/properties/src/provider/bidi_data.rs [hunk 17]
```diff
@@ -202,16 +174,15 @@ impl AsULE for MirroredPairedBracketData {
     fn from_unaligned(unaligned: Self::ULE) -> Self {
         // Safe because the lower bits 20..0 of MirroredPairedBracketDataULE bytes are the CharULE bytes,
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:175]: Nit: Move the safety comment to the line directly before the `unsafe`

File: components/properties/src/trievalue.rs (hunk 14 above)
Comment (sffc) [reply components/properties/src/trievalue.rs:180]: I think what you did here is good; map to `ZeroVecError` when you need to, and otherwise stick with a more expressive error type.

I observe that this function allows all values of the open enum; `BidiPairedBracketType` takes 0, 1, 2, or 3. I have another thread for discussing whether that is good or bad or neutral.

File: components/properties/src/provider/bidi_data.rs [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+            let mut mirroring_glyph_code_point: u32 = (byte2 & 0x1F) as u32;
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte1 as u32);
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte0 as u32);
+            let _mirroring_glyph = char::from_u32(mirroring_glyph_code_point)
+                .ok_or(ZeroVecError::VarZeroVecFormatError)?;
+
+            // skip validating the Bidi_Mirrored boolean since it is always valid
+
+            // assert that Bidi_Paired_Bracket_Type cannot have a 4th value because it only
+            // has 3 values: Open, Close, None
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/properties/src/provider/bidi_data.rs:145]: I'll strengthen my Thought to an Issue since there are several ways to construct a MirroredPairedBracketDataULE with a value of 3 for Bidi_Paired_Bracket_Type and we shouldn't impose more strict requirements on the deserialization of it.

All you need to do is remove this check. The only restriction on the ULE needs to be that the bottom 21 bits are a valid code point.

[the remark this answers is not in the document]
Comment (sffc) [reply components/properties/src/provider/bidi_data.rs:145]: Suggestion: Change the whole contents of this loop to

```rust
    let [byte0, byte1, byte2] = *<&[u8; 3]>::try_from(byte_triple).unwrap();
    let u = u32::from_le_bytes([byte0, byte1, byte2 & 0x1F, 0]);
    char::try_from(u).map_err(|_| ZeroVecError::parse::<Self>())?;
```

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/properties/src/provider/bidi_data.rs:145]: Yeah, we should either make this an open integer type, or close the enum.

Note that the discriminant here does not have the full u8 range.

File: components/properties/src/provider/bidi_data.rs [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+        let mut byte2 = byte_slice.get(2).copied().unwrap_or_default();
+        byte2 |= (self.mirrored as u8) << 5;
+        byte2 |= self.paired_bracket_type.0 << 6;
+
+        MirroredPairedBracketDataULE([
+            byte_slice.first().copied().unwrap_or_default(),
+            byte_slice.get(1).copied().unwrap_or_default(),
+            byte2,
+        ])
+    }
```
Comment (sffc) [inline components/properties/src/provider/bidi_data.rs:171]: Suggestion: re-write this method as

```suggestion
    fn to_unaligned(self) -> Self::ULE {
        let ch = u32::from(self.mirroring_glyph);
        ch |= u32::from(self.mirrored) << 21;
        ch |= self.paired_bracket_type.0 << 22;
        let [byte0, byte1, byte2, _] = ch.to_le_bytes();
        MirroredPairedBracketDataULE([byte0, byte1, byte2])
    }
```

File: components/properties/src/provider/bidi_data.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (echeran) [reply components/properties/src/provider/bidi_data.rs:145]: @Manishearth This was the check that I was skipping (b/c the enum is open) that we discussed together and changed as a result. I think you said that Bidi_Paired_Bracket_Type will not add a property value beyond Open/Close/None, so we should have a stronger assertion accordingly, regardless of whether the Rust enum is open or not.

Of course, the only enumerated property guaranteed by the spec to be closed is General_Category. Everything else is theoretically subject to adding new property values.

I'm not terribly opinionated here because my intuition is weak, but I want to make sure that we can explain why we choose whichever decision we make here.

File: provider/testdata/src/lib.rs [hunk 20]
```diff
@@ -92,7 +92,7 @@ pub mod versions {
     /// # Examples
     ///
     /// ```
-    /// assert_eq!("icu4x/2023-02-24/72.x", icu_testdata::versions::icu_tag());
+    /// assert_eq!("icu4x/2023-03-22a/72.x", icu_testdata::versions::icu_tag());
```
Comment (sffc) [inline provider/testdata/src/lib.rs:95]: Nit: I'd prefer for the tag change to be its own standalone PR (I've given similar feedback to others in the past)

File: components/properties/src/provider/bidi_data.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (echeran) [reply components/properties/src/provider/bidi_data.rs:211]: Talking with @Manishearth, he reminded me that one of the ULE invariants that has to be enforced is that we cannot have a ULE byte serialization that corresponds to invalid data in its original form.

That hasn't been a problem for our property enums because they are all open enums and they all serialize to their own separate ULE type. Anything is valid in both the enum discriminant and the byte serialization.

Here, we're doing this bit packing for MirroredPairedBracketData's ULE type, and we've only allocated 2 bits to represent Bidi_Paired_Bracket_Type. So we're assuming that no new property values are defined (if they did, they would show up in a pair, probably, and we wouldn't be able to represent 3+2 = 5 property values in the 2 bits we've allocated in the bit packing). So we definitely aren't treating this as an open enum, so we have to treat it as a closed one.

I started down the path of the design using a closed version of the enum in MirroredPairedBracketData, and making it private, in to enforce the invariant while also avoiding having to write a custom impl of DataBake, but I can't tell from what I've done if I've misunderstood something in my new code or if it's possible to avoid writing a custom impl of DataBake.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/properties/src/provider/bidi_data.rs:211]: >  and making it private

we shouldn't make provider module types private

> in to enforce the invariant while also avoiding having to write a custom impl of DataBake, but I can't tell from what I've done if I've misunderstood something in my new code or if it's possible to avoid writing a custom impl of DataBake

What's wrong with slapping a `#[derive(Bake)]` onto it?

What you need here is your BPTClosed enum with Bake/Serialize/Deserialize custom derive impls on it, and `TryFrom<BPT>` (and `From` in the other direction)

File: provider/testdata/src/lib.rs (hunk 20 above)
Comment (echeran) [reply provider/testdata/src/lib.rs]: I just remembered: in order for `ci-job-testdata` to pass, the tag needs to be updated because it refers to the data from the latest tag when running `cargo make testdata`. So if we want the latest version tag to be changed in a separate PR, then merging this PR will require overriding the fact that the relevant CI check will be failing (and that the follow on CI check `full-datagen` won't run as a result, too).

Comment (sffc) [reply provider/testdata/src/lib.rs]: You can include the tag update in this PR but also pull it out, merge the other PR, and then update this PR to the latest main, which should be a small change. But I'll leave it up to you if you want to do this. It's a preference of mine but not super strong since this is only an additive change.

File: components/properties/src/bidi_data.rs [hunk 21]
```diff
@@ -154,8 +155,8 @@ impl<'a> BidiAuxiliaryPropertiesBorrowed<'a> {
         let mirroring_glyph = bidi_aux_props.mirroring_glyph;
         let paired_bracket_type = bidi_aux_props.paired_bracket_type;
         match paired_bracket_type {
-            BidiPairedBracketType::Open => BidiPairingProperties::Open(mirroring_glyph),
-            BidiPairedBracketType::Close => BidiPairingProperties::Close(mirroring_glyph),
+            CheckedBidiPairedBracketType::Open => BidiPairingProperties::Open(mirroring_glyph),
+            CheckedBidiPairedBracketType::Close => BidiPairingProperties::Close(mirroring_glyph),
```
Comment (sffc) [inline components/properties/src/bidi_data.rs:159]: I think this is the place where you consume the `CheckedBidiPairedBracketType`, right? I don't understand the advantage of using the closed enum instead of using the open enum throughout.

File: components/properties/src/provider/bidi_data.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (sffc) [reply components/properties/src/provider/bidi_data.rs:204]: I don't think we need both the open enum and the closed enum. One should be sufficient.

I also don't understand why the open enum is violating a ULE invariant. I mean, if you made a `BMBT(45)`, it gets serialized as `0b01` and then comes back as `BMBT(1)`, which is a little weird but is it an invariant? It is the equality invariant at best, which could in principle be solved by a custom `PartialEq` on the open enum. Right?

All that said, I'm perfectly fine with a closed enum here. We should be able to use it throughout. We can add back the `(byte3 & 0xC0) != 0xC0` check and then we're good to go.

File: components/properties/src/bidi_data.rs (hunk 21 above)
Comment (Manishearth) [reply components/properties/src/bidi_data.rs:159]: We can't use the open enum throughout: it has states that cannot be represented in the ULE type because we don't have space.

If we want to use the open enum, `MirroredPairedBracketData` needs to have private fields, with a custom validating serde impl (and a custom ctor for databake).

File: components/properties/src/bidi_data.rs [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+    /// assert_eq!(open_angle_bracket.mirroring_glyph, Some('>'));
+    /// assert_eq!(open_angle_bracket.mirrored, true);
+    /// let close_angle_bracket = bidi_data.get32_mirroring_props('>' as u32);
+    /// assert_eq!(close_angle_bracket.mirroring_glyph, Some('<'));
+    /// assert_eq!(close_angle_bracket.mirrored, true);
+    /// let three = bidi_data.get32_mirroring_props('3' as u32);
+    /// assert_eq!(three.mirroring_glyph, None);
+    /// assert_eq!(three.mirrored, false);
+    /// ```
+    pub fn get32_mirroring_props(&self, code_point: u32) -> BidiMirroringProperties {
```
Comment (Manishearth) [inline components/properties/src/bidi_data.rs:119]: suggestion (nb): If we do not plan to have char-accepting methods, do we need to call this `get32`?

(There's not much point to having char-accepting methods since they won't be faster and the caller can `as u32` when needed)

File: components/properties/src/props.rs (hunk 13 above)
Comment (Manishearth) [reply components/properties/src/props.rs:302]: Yeah, it's necessary to be able to make CPTs for datagen. That said, we _could_ `cfg_attr(feature = "datagen)` this, or just use `CPT<u8>` in datagen.

File: components/properties/src/props.rs [hunk 23]
```diff
@@ -284,6 +289,29 @@ impl BidiClass {
...
+pub struct BidiPairedBracketType(pub u8);
+
+#[allow(non_upper_case_globals)]
+impl BidiPairedBracketType {
```
Comment (Manishearth) [inline components/properties/src/props.rs:306]: thought: hmm, so I was under the impression we already had this type when I suggested making a closed and open.

If this is a new type, perhaps this can just be a closed enum for now (and we have a single type). @sffc thoughts?

Comment (Manishearth) [reply components/properties/src/props.rs:306]: this is being discussed below

File: components/properties/src/provider/bidi_data.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (echeran) [reply components/properties/src/provider/bidi_data.rs:204]: Ok, and just so I understand, the reason we are okay to get rid of the open enum `BidiPairedBracketType` is because we don't have any use cases currently to support just the enum on its own (ex: no ECMA-402 requirement), but instead our only use case is via this space efficient Bidi properties struct?

I can get rid of `BidiPairedBracketType` and just have the `Checked-` version, as described. I'd like to keep the names unchanged so that if we ever need to faithfully represent Bidi_Paired_Bracket_Type property values as they are, we can revive the names without confusion.

[the remark this answers is not in the document]
Comment (sffc) [reply components/properties/src/provider/bidi_data.rs:204]: SGTM. Name doesn't matter too much since it is only a data provider type not directly exposed through the primary API

File: components/properties/src/bidi_data.rs (hunk 22 above)
Comment (sffc) [reply components/properties/src/bidi_data.rs:123]: I suggested get32 because I think we should also have char methods

File: components/properties/src/trievalue.rs (hunk 15 above)
Comment (sffc) [reply components/properties/src/trievalue.rs:102]: Well, now that we're getting strict again with this being a closed enum, how about you move this type to be defined directly in the datagen crate, if that is the only user?

File: provider/testdata/src/lib.rs (hunk 20 above)
Comment (echeran) [reply provider/testdata/src/lib.rs]: Done.

File: components/properties/src/trievalue.rs (hunk 15 above)
Comment (echeran) [reply components/properties/src/trievalue.rs]: Inside the implementation of one of the APIs in the properties crate, we need to match the `CheckedBidiPairedBracketType` enum value in order to return the correct value for the struct.  (`BidiAuxiliaryPropertiesBorrowed::get32_pairing_props`)

File: components/properties/src/error.rs [hunk 24]
```diff
@@ -31,6 +31,9 @@ pub enum PropertiesError {
     /// An unknown or unexpected property name was used for an API dealing with properties specified as strings at runtime
     #[displaydoc("Unexpected or unknown property name")]
     UnexpectedPropertyName,
+    /// A `u32` serialized value of `MirroredPairedBracketData` did not encode either a valid Bidi_Mirroring_Glyph or a valid Bidi_Paired_Bracket_Type
+    #[displaydoc("Invalid MirroredPairedBracketData serialized in int: {0}")]
+    MirroredPairedBracketDataFromU32(u32),
```
Comment (sffc) [inline components/properties/src/error.rs:36]: Suggestion: I think slightly more idiomatic in Rust would be for this to be its own little error struct named `MirroredPairedBracketDataTryFromError`

**Review by sffc — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]
